# SNP-pipeline
Underlying biology and problem description: We are living in the post-genomic era where sequencing genomes now takes under a day; thousands of genomes have  been  sequenced  and  assembled.  Many  (not  all)  organisms,  whether prokaryote  or  eukaryote,  now  have  a  complete  reference  (or  a  canonical) genome sequenced and assembled already. Given that assembling genomes de novo requires substantial expertise along with computational and human resources,  a  more  practical  approach  of  analyzing  genomes  is  to  map  the genomic reads to an existing reference. This approach is also referred to as genome re-sequencing. Once you have your reads mapped to an existing reference strain, the process can diverge into multiple directions depending on the nature and objective of the project. One common direction is to call variants. Variants are simply the bases  in  your sampleâ€™s  genome  that  are  different  from  the  corresponding bases in the reference genome. These differences can (or cannot) lead to a range of phenotypic differences either directly (as is the case in amino acid changes  or  nonsynonymous  changes)  or  indirectly  (a  change  of  base  in  the splice  site  leading  to  change  in  splicing  pattern  or  change  in  the  regulatory region leading to enhanced or depleted gene expression). 
The pipeline will have the following steps:
1. Align FASTQ reads to a reference genome to create an alignment file using BWA-MEM
2. Processing  the  alignment  file  (file  format  conversion,  sorting,  alignment improvement) using Samtools
3. Calling the variants using GATK (Genome Analysis Tool Kit)
